Move RuntimeSettings into framework, build modules as binaries

hikinggrass opened the pull request:
Modules can still be built and loaded as shared libraries by setting the CMake option BUILD_MODULE_SHARED_LIBRARIES to ON

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

a-w50 commented on CMakeLists.txt:
```diff
@@ -37,6 +37,10 @@ endif()
 
+option(BUILD_MODULE_SHARED_LIBRARIES "Build modules as shared libraries" OFF)
+
+option(BUILD_MODULE_BINARIES "Build modules as binaries" ON)
```
Is there any reason why we should allow both options instead of xor?
Do shared modules have any advantage over the binaries?

a-w50 commented on include/framework/runtime.hpp:
```diff
@@ -0,0 +1,80 @@
+    std::function<void(ModuleAdapter::ExtMqttSubscribeFunc ext_mqtt_subscribe)> register_ext_mqtt_subscribe_handler;
+    std::function<std::vector<cmd>(const json& connections)> everest_register;
+    std::function<void(ModuleConfigs module_configs, const ModuleInfo& info)> init;
+    std::function<void()> ready;
```
Probably this commit would be a good point to remove all those different register_handler functions.  Except for the ext_mqtt functions we want to register always all the functions and not just one by one.  I would propose, that instead there will be only one single hook, which the module as a library (in case we'll keep it as a library) will offer - something like register_everest, and this single function will then receive a struct with all the necessary functions to access everest.  So instead of registering n single callbacks, one register with a struct of n callbacks.

hikinggrass replied:
Sure, that sounds good. I'll do a little more refactoring in this series of PRs.

a-w50 commented on lib/runtime.cpp:
```diff
@@ -0,0 +1,228 @@
+    everest_register(everest_register),
+    init(init),
+    ready(ready) {
+}
```
This could be avoided by the above mentioned single register function

a-w50 commented on lib/runtime.cpp:
```diff
@@ -0,0 +1,228 @@
+        everest.signal_ready();
+
+        mainloop_thread.join();
+
```
I guess there is a lot of code duplication here with main.cpp.  If we will drop main.cpp, then this doesn't matter.  If not, it would be nice to share common source.

hikinggrass replied:
I think I'll drop the modules as shared libraries completely and remove the main.cpp in that process.

a-w50 commented on lib/runtime.cpp:
```diff
@@ -0,0 +1,228 @@
+    return 0;
+}
+
+bool ModuleLoader::parse_command_line(int argc, char* argv[]) {
+    po::options_description desc("EVerest");
+    desc.add_options()("help,h", "produce help message");
+    desc.add_options()("main_dir", po::value<std::string>(), "Set main EVerest directory");
+    desc.add_options()("module,m", po::value<std::string>(),
+                       "Which module should be executed (module id from config file)");
+    desc.add_options()("log_conf", po::value<std::string>(), "The path to a custom logging.ini");
+    desc.add_options()("dontvalidateschema", "Don't validate json schema on every message");
+    desc.add_options()("conf", po::value<std::string>(), "The path to a custom config.json");
+
+    po::variables_map vm;
+    po::store(po::parse_command_line(argc, argv, desc), vm);
+    po::notify(vm);
+
+    if (vm.count("help") != 0) {
+        std::cout << desc << "\n";
+        return false;
+    }
+
+    if (vm.count("main_dir") == 0) {
+        EVTHROW(EVEXCEPTION(EverestApiError, "--main_dir parameter is required"));
+    }
+
+    this->runtime_settings = std::make_unique<RuntimeSettings>(vm);
+    this->original_process_name = argv[0];
+
+    if (vm.count("module") != 0) {
+        this->module_id = vm["module"].as<std::string>();
+    } else {
+        EVTHROW(EVEXCEPTION(EverestApiError, "--module parameter is required"));
+    }
+
+    return true;
```
Although this gives additional flexibility, I don't really like the fact, that the manager has already parsed all the arguments, just to forward them and let them parsed again.  Something related to this would also be, that the spawned modules shouldn't know about the complete configuration - shouldn't they only get their module and implementation specific config (which could be passed by the manager via mqtt?)

hikinggrass replied:
I like the idea of removing the double-parsing of the config a lot, but since this will probably be a bigger change I'll tackle that in a dedicated PR

a-w50 reviewed (approved):
I'm happy with the conceptual implementation, although I think we could do more refactoring and unification.
The difference between the shared modules and the binary is in the end just a matter of linkage.  Concerning the capability stuff we could equally put the same main binary into each of the module path and let it load the shared library module.
But I would almost argue, that we don't need the library as a shared module, as long as we are using it in a separate process, because we win nothing using it - i.e. we want to have the capabilty feature and hence need a separate loader binary for every module, the shared module library doesn't get used by multiple processes, so no memory advantages, and lastly the dynamic loading is less performant than a statically linked binary (that might be still link dynamically against libeverest).

hikinggrass committed:
Move RuntimeSettings into framework, build modules as binaries by default

Modules can still be built and loaded as shared libraries by setting the CMake option BUILD_MODULE_SHARED_LIBRARIES to ON

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

# Conflicts:
#	src/manager.cpp

hikinggrass committed:
Remove main binary as shared library modules are no more

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Replace register_* functions with register_module_adapter

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>
